Magic smpi new

hb326 opened the pull request:
Calculating the SMPI with nine out of the original twelve variables is fine. Provenance is implemented and the documentation is added. 
The last three variables cannot be added just yet since there is still a problem in the pre-processing of the model files. The code will be updated as soon as the problem is solved.

bouweandela commented on esmvaltool/config-user_MISTRAL.yml:
```diff
@@ -0,0 +1,26 @@
+# ESMValTool User's configuration file for DKRZ-MISTRAL
```
Was this accidentally committed?

hb326 replied:
Yes, that was an accident. I removed the file now.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,125 @@
+
+#. main.ncl
+
+   *Required diag_script_info attributes*
```
```suggestion
   *Required settings for script*
```

hb326 replied:
Changed.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,125 @@
+   * time_avg: type of time average (opt argument of time_operations in diag_scripts/shared/statistics.ncl)
+   * region: selected region (see select_region in diag_scripts/shared/latlon.ncl)
+   
+   *Optional diag_script_info attributes*
```
```suggestion
   *Optional settings for script*
```

hb326 replied:
Changed.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,125 @@
+   * latlon_cmap: for plot_type latlon only, chosen color table (default: "amwg_blueyellowred")
+   * plot_units: plotting units (if different from standard CMOR units)
+   
+   *Required variable_info attributes*
```
```suggestion
   *Required settings for variables*
```

hb326 replied:
Changed.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,125 @@
+   
+   * reference_dataset: reference dataset to compare with (usually the observations).
+   
+   *Optional variable_info attributes*
```
```suggestion
   *Optional settings for variables*
```

hb326 replied:
Changed.

hb326 commented:
> 
> 
> Can all possible settings mentioned in the documentation actually work for this particular recipe?

The SMPI is very specific in what in needs to be calculated. So there is not much choice on what can and should be used as settings. However, the program "main.ncl" has the option for all the settings mentioned, and that's why I listed them here. Do you think I should only list the settings that we use in the SMPI?

bouweandela commented:
> Do you think I should only list the settings that we use in the SMPI?

Yes, I think the recipe page should list only the settings that can be changed by a recipe user without breaking the recipe. Maybe it would even be good to mention that any other settings should not be changed.

bouweandela commented:
I think we could also try to merge this with just 9 variables and then add the 3 missing ones later in a new pull request. That way we could probably get this running before the final MAGIC deadline at the end of this week.

I just tested this on Jasmin, and it runs mostly fine, but it fails at the collect script:
```
INFO    Loading settings from /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190304_134637/run/collect/SMPI/settings.ncl
fatal:Number of subscripts do not match number of dimensions of variable,(3) Subscripts used, (2) Subscripts expected
^Mfatal:["Execute.c":8637]:Execute: Error occurred at or near line 156 in file /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/ESMValTool/esmvaltool/diag_scripts/perfmetrics/collect.ncl
```

hb326 commented:
> Yes, I think the recipe page should list only the settings that can be changed by a recipe user without breaking the recipe. Maybe it would even be good to mention that any other settings should not be changed.

Ok, I will adjust the description then.

hb326 commented:
> 
> 
> I think we could also try to merge this with just 9 variables and then add the 3 missing ones later in a new pull request. That way we could probably get this running before the final MAGIC deadline at the end of this week.
> 
> I just tested this on Jasmin, and it runs mostly fine, but it fails at the collect script:
> 
> ```
> INFO    Loading settings from /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190304_134637/run/collect/SMPI/settings.ncl
> fatal:Number of subscripts do not match number of dimensions of variable,(3) Subscripts used, (2) Subscripts expected
> ^Mfatal:["Execute.c":8637]:Execute: Error occurred at or near line 156 in file /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/ESMValTool/esmvaltool/diag_scripts/perfmetrics/collect.ncl
> ```

I am trying to reproduce the problem that you had. When I run the recipe as it is present in the MAGIC_SMPI_new branch I do not get this error. Did you use the recipe as it is in the branch or did you adjust anything?

bouweandela commented:
I'm using the recipe as is, but skipping datasets that are not available in the CP4CDS collection with the `--skip-nonexistent` flag.

hb326 commented:
> 
> 
> I think we could also try to merge this with just 9 variables and then add the 3 missing ones later in a new pull request. That way we could probably get this running before the final MAGIC deadline at the end of this week.
> 
> I just tested this on Jasmin, and it runs mostly fine, but it fails at the collect script:
> 
> ```
> INFO    Loading settings from /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190304_134637/run/collect/SMPI/settings.ncl
> fatal:Number of subscripts do not match number of dimensions of variable,(3) Subscripts used, (2) Subscripts expected
> ^Mfatal:["Execute.c":8637]:Execute: Error occurred at or near line 156 in file /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/ESMValTool/esmvaltool/diag_scripts/perfmetrics/collect.ncl
> ```

We adapted "collect.ncl". Hopefully that will fix the problem.
However, it seems like the problem occured since there are not all variables available for the two models. Is that correct? For the SMIP, all models need all given variables for the exact same time period.

bouweandela commented:
> However,  it seems like the problem occured since there are not all variables  available for the two models. Is that correct? For the SMIP, all models  need all given variables for the exact same time period.

Yes, this was indeed the case. I've modified the recipe so it contains only the datasets that have all variables available in the CP4CDS datasets.

However, when I try to run the diagnostic now, it crashes because there are syntax errors in interface_scripts/auxiliary.ncl, see the attached log file:
[log.txt](https://github.com/ESMValGroup/ESMValTool/files/2941420/log.txt)

hb326 commented:
> 
> 
> > However,  it seems like the problem occured since there are not all variables  available for the two models. Is that correct? For the SMIP, all models  need all given variables for the exact same time period.
> 
> Yes, this was indeed the case. I've modified the recipe so it contains only the datasets that have all variables available in the CP4CDS datasets. It would be good if you could mention this restriction in the documentation.
> 
> However, when I try to run the diagnostic now, it crashes because there are syntax errors in interface_scripts/auxiliary.ncl, see the attached log file: [log.txt](https://github.com/ESMValGroup/ESMValTool/files/2941420/log.txt)
> This is rather strange, since that file appears not to have been modified in the pull request. Can you try to find out what is going on?

Will try. It might be an update from the branch "version2_development" that is interfering with something...

hb326 commented:
> 
> 
> > > However,  it seems like the problem occured since there are not all variables  available for the two models. Is that correct? For the SMIP, all models  need all given variables for the exact same time period.
> > 
> > 
> > Yes, this was indeed the case. I've modified the recipe so it contains only the datasets that have all variables available in the CP4CDS datasets. It would be good if you could mention this restriction in the documentation.
> > However, when I try to run the diagnostic now, it crashes because there are syntax errors in interface_scripts/auxiliary.ncl, see the attached log file: [log.txt](https://github.com/ESMValGroup/ESMValTool/files/2941420/log.txt)
> > This is rather strange, since that file appears not to have been modified in the pull request. Can you try to find out what is going on?
> 
> Will try. It might be an update from the branch "version2_development" that is interfering with something...

@bouweandela: I tried all different kinds of things to see where the problem could originate from. But so far I could not reproduce it. Even with merging the latest version of "version2_development" into my local "MAGIC_SMPI_new" branch. 
The log file indicates that the diagnostic had a problem with the "pr" variable. Have you tried to run the diagnostic without "pr"? Maybe it is a corrupt "pr" file?
And maybe @mattiarighi has another idea what could be the problem?

hb326 commented:
> 
> 
> > Do you think I should only list the settings that we use in the SMPI?
> 
> Yes, I think the recipe page should list only the settings that can be changed by a recipe user without breaking the recipe. Maybe it would even be good to mention that any other settings should not be changed.

@bouweandela: I updated the description of the diagnostics, and added an explanation for the fixed settings. Let me know if the changes do not make sense.

bouweandela commented:
Thanks for testing. I will try to run it again, maybe it's some issue with the Jasmin cluster. Performance has been quite flaky all day.

bouweandela commented:
I've run it again on Jasmin, the perviously reported error did not appear. However, the grading script crashes with the following log message:
```
Diagnostic script perfmetrics/collect.ncl failed with return code -11.
```
and the NCL log message is
```
INFO    Loading settings from /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190307_165016/run/collect/SMPI/settings.ncl
schleife 1
schleife 1
schleife 1
schleife 1
schleife 1
schleife 1
schleife 1
schleife 1
schleife 1
INFO    Wrote variable SMPI to /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190307_165016/work/collect/SMPI//_SMPI.nc
warning:LlDataLineTo: point 0.000000,0.850000 outside data domain
^Mwarning:LlDataLineTo: point 0.000000,1.150000 outside data domain
```

hb326 commented:
> 
> 
> I've run it again on Jasmin, the perviously reported error did not appear. However, the grading script crashes with the following log message:
> 
> ```
> Diagnostic script perfmetrics/collect.ncl failed with return code -11.
> ```
> 
> and the NCL log message is
> 
> ```
> INFO    Loading settings from /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190307_165016/run/collect/SMPI/settings.ncl
> schleife 1
> schleife 1
> schleife 1
> schleife 1
> schleife 1
> schleife 1
> schleife 1
> schleife 1
> schleife 1
> INFO    Wrote variable SMPI to /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/MAGIC_SMPI_new/output_dir/recipe_smpi_20190307_165016/work/collect/SMPI//_SMPI.nc
> warning:LlDataLineTo: point 0.000000,0.850000 outside data domain
> ^Mwarning:LlDataLineTo: point 0.000000,1.150000 outside data domain
> ```

Will look into this. Unfortunately, I will only be able to do that on Monday.

bouweandela commented:
Note that I updated the issue. It looks like the crash occurs in the provenance logging.

bouweandela commented:
This was apparently fixed in https://github.com/ESMValGroup/ESMValTool/pull/935, so merging the latest version of the development branch fixed this issue.

bouweandela commented on esmvaltool/diag_scripts/perfmetrics/collect.ncl:
```diff
@@ -204,12 +396,39 @@ begin
+  ; collect meta-data and call ESMValMD function
+  nc_file = ncdf_outfile
+  plot_file = outfile_plot
+  domain = ("glob")
```
```suggestion
  domains = ("global")
```

hb326 replied:
Ok. I will keep that in mind.
The changes are ok with me.

bouweandela commented on esmvaltool/diag_scripts/perfmetrics/collect.ncl:
```diff
@@ -204,12 +396,39 @@ begin
+    references = (/"righi15gmd", "gleckler08jgr"/)
+    infiles = "will still have to be defined!"     ; TO FIX!
+  end if
+  log_provenance(nc_file, plot_file, caption, statistics, domain, plottype, \
```
```suggestion
  log_provenance(nc_file, plot_file, caption, statistics, domains, plottype, \
```
see #951

hb326 replied:
Ok.
These changes also look good to me.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+User settings
+-------------
+
+#. main.ncl
```
```suggestion
#. perfmetrics/main.ncl
```
path should be relative to esmvaltool/diag_scripts

bouweandela commented:
@hb326 @bjoernbroetz Would it be possible to finalize this pull request? We would really like to finish all Magic diagnostics as soon as possible.

I think the following issues should be addressed before this can be merged:
- Recipe documentation should be updated so it only lists settings that can be changed and mentions that other settings should not be changed
- Style issues need to be fixed so the unit tests pass
- Merge conflict should be resolved
- This recipe shares a lot of code with perfmetrics. Wouldn't it make more sense to move the shared code to the esmvaltool/diag_scripts/shared directory? Having code specific to the smpi recipe in the esmvaltool/diag_scripts/perfmetrics directory is confusing too, so I would move that to a esmvaltool/diag_scripts/smpi directory.

mattiarighi commented:
> This recipe shares a lot of code with perfmetrics. Wouldn't it make more sense to move the shared code to the esmvaltool/diag_scripts/shared directory? Having code specific to the smpi recipe in the esmvaltool/diag_scripts/perfmetrics directory is confusing too, so I would move that to a esmvaltool/diag_scripts/smpi directory.

smpi simply extends the perfmetrics existing code with an additional metric.
Therefore I think the scripts should stay in `perfmetrics/`.

bouweandela commented:
> smpi simply extends the perfmetrics existing code with an additional metric.
Therefore I think the scripts should stay in perfmetrics/.

Does that mean that smpi metric will also become part of the perfmetrics recipe?

mattiarighi commented:
No, but different recipes can use the same diagnostics.
This happens for example for some of the diagnostics in `recipe_flato13ipcc.yml`.
There is no need to have a 1-to-1 correspondance between recipes and diagnostics.

bouweandela commented:
Yes, multiple recipes can use the same diagnostic script, if they use it in the same way.

If you're going to implement 'special case if called with this or that option', and that option belongs to a very specific diagnostic, as is done here, things are rapidly becoming difficult to follow and test: adding these special cases to the perfmetrics code, means that to test this pull request, you will also have to  re-test the perfmetrics recipe and the flato13ipcc recipe and you have to know that these other recipes are using the same code, because you cannot see it from the path name.

My advice would be to put special case code should be in a file specific to the diagnostic it belongs to and general code in the shared directory. Of course it's up to you if you want to follow that advice or not.

mattiarighi commented:
I understand what you mean, but it is not so easy to do that without a deep restructuring of perfmetrics (which we want to avoid at this stage).

smpi is just an additional metric (like the existing RMSD or bias), which to be applied requires some changes to the existing code.

I think the easiest solution here is to modify the existing code (as @hb326 did) and, of course, test all the recipes which use it to make sure it still runs properly. That's anyway less work that restructuring everything (again).

mattiarighi commented:
@hb326 if you want to proceed the way @bouweandela suggests, we can briefly discuss that.

hb326 commented:
>     * Recipe documentation should be updated so it only lists settings that can be changed and mentions that other settings should not be changed

I have updated the description about 20 days ago along those lines. Do you think it is not clear enough?

hb326 commented:
>     * #897 is now closed, so I think the additional variables can also be added

Yes, I had planned on doing that soon. I will do that sooner now.

hb326 commented:
> 
> 
> @hb326 if you want to proceed the way @bouweandela suggests, we can briefly discuss that.

@mattiarighi, let's talk about this when you are back on Friday.

bouweandela commented:
> I understand what you mean, but it is not so easy to do that without a deep restructuring of perfmetrics (which we want to avoid at this stage).
>
> smpi is just an additional metric (like the existing RMSD or bias), which to be applied requires some changes to the existing code.
>
> I think the easiest solution here is to modify the existing code (as @hb326 did) and, of course, test all the recipes which use it to make sure it still runs properly. That's anyway less work that restructuring everything (again).

It really depends a on what your plans for the future of the perfmetrics code are. If this is the last addition, it's probably not worth the effort to 'do the right thing'. Conversely, if you plan to build more diagnostics that use the perfmetrics code, I think it is better to refactor now, because now you know exactly what hte SMPI code does since you've just written it, so that leaves just 2 metrics to disentangle. The longer you wait with refactoring the more work it will become to separate the code and test.

bouweandela commented:
> I have updated the description about 20 days ago along those lines. Do you think it is not clear enough?

@hb326 My apologies, I hadn't noticed that you changed it. I will test the available options and review it again.

mattiarighi commented:
> It really depends a on what your plans for the future of the perfmetrics code are. If this is the last addition, it's probably not worth the effort to 'do the right thing'. Conversely, if you plan to build more diagnostics that use the perfmetrics code, I think it is better to refactor now, because now you know exactly what hte SMPI code does since you've just written it, so that leaves just 2 metrics to disentangle. The longer you wait with refactoring the more work it will become to separate the code and test new stuff.

I don't have any further plans with perfmetrics at the moment, but who knows what other users may need. Let me have a closer look with @hb326 on Friday...

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+   *Required settings for script*
+
+   * plot_type: cycle (time), zonal (plev, lat), latlon (lat, lon), cycle_latlon (time, lat, lon)
+   * time_avg: type of time average (opt argument of time_operations in diag_scripts/shared/statistics.ncl)
```
Either list the available options here, or solve #809 and add a link to the documentation of that function here, so users can click it. Expecting users to go and read your source code is nowhere near user friendly.

hb326 replied:
I have adopted this description directly from perfmetrics, so that the descriptions are consistent.

bouweandela replied:
The documentation of perfmetrics should also be fixed, that's why I opened issue #809 in the first place.

mattiarighi replied:
Agree, but better in another PR.

bouweandela replied:
That's fine, but the documentation for the magic project should be ready soon. Who is going to fix that issue and when?

mattiarighi replied:
The issue was originally assigned to @axel-lauer because he took care of this in version 1.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+
+   * plot_type: cycle (time), zonal (plev, lat), latlon (lat, lon), cycle_latlon (time, lat, lon)
+   * time_avg: type of time average (opt argument of time_operations in diag_scripts/shared/statistics.ncl)
+   * region: selected region (see select_region in diag_scripts/shared/latlon.ncl)
```
Same here (even worse, since the available options are not even in the documentation of the function)

hb326 replied:
Same here: directly adopted from perfmetrics.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+   * plot_type: cycle (time), zonal (plev, lat), latlon (lat, lon), cycle_latlon (time, lat, lon)
+   * time_avg: type of time average (opt argument of time_operations in diag_scripts/shared/statistics.ncl)
+   * region: selected region (see select_region in diag_scripts/shared/latlon.ncl)
+   * normalization: metric normalization (for RMSD and BIAS metrics only)
```
This does not appear to apply to SMPI?

hb326 replied:
Good point. I have fixed this.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+   * time_avg: type of time average (opt argument of time_operations in diag_scripts/shared/statistics.ncl)
+   * region: selected region (see select_region in diag_scripts/shared/latlon.ncl)
+   * normalization: metric normalization (for RMSD and BIAS metrics only)
+   * calc_grading: calculates grading metrics (default: False; has to be set to "true")
```
I would remove the default value, this is confusing: are you referring to the default value in the recipe or in the diagnostic script?

hb326 replied:
I have adjusted this.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+   * calc_grading: calculates grading metrics (default: False; has to be set to "true")
+   * metric: chosen grading metric(s) (if calc_grading is True; has to be set to "SMPI")
+   * smpi_n_bootstrap: number of bootstrapping members used to determine uncertainties on model-reference differences
+   
```
In the recipe there is an undocumented `range_option`, what is that about?

hb326 replied:
That is a remnant from an older version of perfmetrics v2. I removed it from the recipe.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+   * normalization: metric normalization (for RMSD and BIAS metrics only)
+   * calc_grading: calculates grading metrics (default: False; has to be set to "true")
+   * metric: chosen grading metric(s) (if calc_grading is True; has to be set to "SMPI")
+   * smpi_n_bootstrap: number of bootstrapping members used to determine uncertainties on model-reference differences
```
It would be nice if you could add a range of typical values

hb326 replied:
I have fixed this.

bouweandela commented on doc/sphinx/source/recipes/recipe_smpi.rst:
```diff
@@ -0,0 +1,93 @@
+   
+   *Optional settings for script*
+  
+   * draw_plots: draws plots (default: True)
```
this is set to `false` in the recipe, yet here it says default True, that's confusing

hb326 replied:
I adjusted this in the description.

hb326 commented:
> 
> 
> > ```
> > * #897 is now closed, so I think the additional variables can also be added
> > ```
> 
> Yes, I had planned on doing that soon. I will do that sooner now.

@bouweandela: I have run the recipe in different variations with the three so far not included variables added in. It worked fine now. 
I don't think there is a need for a pull request then, I just updated the recipe on the branch "MAGIC_SMPI_new". Beside uncommenting the three variables, I changed the version for the variable "sic" since that had changed with the update of the observations. I also changed the period for the variable "hfds" since the updated observations cover the whole period now.

mattiarighi commented:
@hb326 I've tested `recipe_perfmetrics_CMIP5.yml` after your changes and, apart a couple of minor issues which I fixed, it works fine and reproduces the results.

Just to be on the safe side, please run again you recipe.

I will finalize the provenance of perfmetrics in a separate PR.

hb326 commented:
> 
> 
> @hb326 I've tested `recipe_perfmetrics_CMIP5.yml` after your changes and, apart a couple of minor issues which I fixed, it works fine and reproduces the results.
> 
> Just to be on the safe side, please run again you recipe.
> 
> I will finalize the provenance of perfmetrics in a separate PR.

I just pulled the updates and ran the recipe again. Works fine for me, looks good, also provenance looks to be still ok.
